src/antsibull/pypi.py

```diff
@@ -0,0 +1,144 @@
+# Copyright (C) 2024 Maxwell G <maxwell@gtmx.me>
+# SPDX-License-Identifier: GPL-3.0-or-later
+# GNU General Public License v3.0+
+# (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+"""
+Work with the PyPI API
+"""
+
+import datetime
+from pathlib import Path
+from typing import NamedTuple, Optional
+from urllib.parse import urljoin
+
+import aiohttp
+import pydantic as p
+from antsibull_core.utils.hashing import verify_hash
+
+PYPI_BASE_URL = "https://pypi.org/pypi/"
+
+
+class SdistAndWheel(NamedTuple):
+    """
+    Tuple of UrlInfo objects for a project's sdist and wheel
+    """
+
+    sdist: "UrlInfo"
+    wheel: "UrlInfo"
+
+
+# pyre-ignore[13]: BaseModel initializes attributes when data is loaded
+class Release(p.BaseModel):
+    """
+    Model representing the response from the PyPI API releases endpoint.
+
+    https://warehouse.pypa.io/api-reference/json.html#get--pypi--project_name---version--json
+    """
+
+    info: "ReleaseInfo"
+    last_serial: int
+    urls: list["UrlInfo"]
+    vulnerabilities: list[dict]
+
+    def get_sdist_and_wheel(self) -> SdistAndWheel:
+        sdist: Optional[UrlInfo] = None
+        wheel: Optional[UrlInfo] = None
+        for release in self.urls:
+            if not sdist and release.packagetype == "sdist":
+                sdist = release
+            elif not wheel and release.packagetype == "bdist_wheel":
+                wheel = release
```

Comment:
Can this else-branch interrupt the loop early?
```suggestion
                wheel = release
            else:
                break
```

Reply:
Is it possible for a `packagetype` that's neither `sdist` nor `bdist_wheel`? I guess not, so this should be a safe optimization. Thanks.

Reply:
Exactly